programs/psa/aead_cipher_psa.c

```diff
@@ -0,0 +1,353 @@
+
+/*
+ * Functions using the PSA Crypto API
+ */
+
+#define CHK( code )     \
+    do {                \
+        status = code;     \
+        if( status != PSA_SUCCESS ) { \
+            printf( "%03d: status = %d\n", __LINE__, status ); \
+            goto exit;  \
+        } \
+    } while( 0 )
+
+static psa_status_t aead_prepare( const char *info,
+                                  psa_key_id_t *key,
+                                  psa_algorithm_t *alg )
+{
+    psa_status_t status;
+
+    size_t key_bits;
+    psa_key_type_t key_type;
+    if( strcmp( info, "gcm128" ) == 0 ) {
+        *alg = PSA_ALG_GCM;
+        key_bits = 128;
+        key_type = PSA_KEY_TYPE_AES;
+    } else if( strcmp( info, "gcm256" ) == 0 ) {
+        *alg = PSA_ALG_GCM;
+        key_bits = 256;
+        key_type = PSA_KEY_TYPE_AES;
+    } else if( strcmp( info, "gcm128_8" ) == 0 ) {
+        *alg = PSA_ALG_AEAD_WITH_SHORTENED_TAG(PSA_ALG_GCM, 8);
+        key_bits = 128;
+        key_type = PSA_KEY_TYPE_AES;
+    } else if( strcmp( info, "chachapoly" ) == 0 ) {
+        *alg = PSA_ALG_CHACHA20_POLY1305;
+        key_bits = 256;
+        key_type = PSA_KEY_TYPE_CHACHA20;
+    } else {
+        puts( usage );
+        return( PSA_ERROR_INVALID_ARGUMENT );
+    }
+
+    psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
+    psa_set_key_usage_flags( &attributes, PSA_KEY_USAGE_ENCRYPT );
+    psa_set_key_algorithm( &attributes, *alg );
+    psa_set_key_type( &attributes, key_type );
+    psa_set_key_bits( &attributes, key_bits );
+
+    CHK( psa_import_key( &attributes, key_bytes, key_bits / 8, key ) );
+
+exit:
+    return( status );
+}
+
+static void aead_info( psa_key_id_t key, psa_algorithm_t alg )
+{
+    psa_key_attributes_t attr = PSA_KEY_ATTRIBUTES_INIT;
+    (void) psa_get_key_attributes( key, &attr );
+    psa_key_type_t key_type = psa_get_key_type( &attr );
+    size_t key_bits = psa_get_key_bits( &attr );
+    psa_algorithm_t base_alg = PSA_ALG_AEAD_WITH_DEFAULT_LENGTH_TAG( alg );
+    size_t tag_len = PSA_AEAD_TAG_LENGTH( key_type, key_bits, alg );
+
+    const char *type_str = key_type == PSA_KEY_TYPE_AES ? "AES"
+                         : key_type == PSA_KEY_TYPE_CHACHA20 ? "Chacha"
+                         : "???";
+    const char *base_str = base_alg == PSA_ALG_GCM ? "GCM"
+                         : base_alg == PSA_ALG_CHACHA20_POLY1305 ? "ChachaPoly"
+                         : "???";
+
+    printf( "aead  : %s, %u, %s, %u\n",
+            type_str, (unsigned) key_bits, base_str, (unsigned) tag_len );
+}
+
+static int aead_encrypt( psa_key_id_t key, psa_algorithm_t alg,
+        const unsigned char *iv, size_t iv_len,
+        const unsigned char *ad, size_t ad_len,
+        const unsigned char *pa, size_t pa_len,
+        const unsigned char *pb, size_t pb_len )
+{
+    psa_status_t status;
+    size_t olen, olen_tag;
+    unsigned char out[32];
+    unsigned char *p = out, *end = out + sizeof( out );
+    unsigned char tag[16];
```

Comment:
Similarly,
```suggestion
    unsigned char tag[PSA_AEAD_TAG_MAX_SIZE];
```
Or better, include the tag with the ciphertext, following RFC 5116.

Reply:
> Or better, include the tag with the ciphertext, following RFC 5116.

That's what I'm doing below, but I'm afraid the multipart API does not make that easy since the `finish()` function requires us to pass a pointer for the tag before it tells us how much plaintext it wrote, so I'm doing this dance with a temporary buffer followed by a `memcpy()`. Did I miss a better way?

Reply:
Btw (loosely related), I'm now wondering if it would be better to demonstrate both the multi-part and the one-shot API. Wdyt?

Reply:
> the finish() function requires us to pass a pointer for the tag before it tells us how much plaintext it wrote

I hadn't thought of that: indeed you need a temporary buffer for the tag. I don't see an easy way to work around that, other than a different finish function that always appends the tag, which is not always desirable. If you want the easy case, there's the one-shot API.

>  I'm now wondering if it would be better to demonstrate both the multi-part and the one-shot API

Sure, we should have demonstrations of the one-shot API. But they're simpler and there's already `docs/getting_started.md` and [Hannes's examples](https://github.com/ARMmbed/mbedtls/pull/5064). I thought that the purpose of your examples was specifically to demonstrate the two multipart APIs side by side.

Reply:
Ok, and thanks for pointing to Hannes's examples, I wasn't aware of them. I was asking because the differences between PSA and legacy are not exactly the same with multi-part and one-shot, but indeed let's avoid redundancy with existing examples.